### PR TITLE
Payable Contracts

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
@canesin NEX should pay attention to this PR.

**Comment by shargon** [comment]
In migrate contract option, we should allow to change dynamic calls and this one too

**Comment by localhuman** [comment]
@shargon I believe you pass in a value for `ContractProperties` which is `HasStorage + HasDynamicInvoke + IsPayable`

Does this ContractProperty change the price of deploy? How is it used?

**Comment by canesin** [comment]
@erikzhang tks for the call, for now is just a new property of contracts, what are your plans for this ?

**Comment by erikzhang** [comment]
> Does this ContractProperty change the price of deploy? How is it used?

No, it doesn't change the price of deploy. It is just a flag for other contracts that want to send tokens into it.

> for now is just a new property of contracts, what are your plans for this ?

I plan to modify the token standards (NEP-5, NFT) so that the contracts must check the `Payable` flag of the receiving contract before transfer, and if not, stop the transfer.

**Comment by localhuman** [comment]
👍 

How will this affect backwards compatibility? I would assume most existing NEP5 contracts would want to be `Payable`.

It seems in this implementation they would default to not payable.

**Comment by erikzhang** [comment]
Most ICO contracts need to be payable. But after the token sale, they will be non-payable. So it doesn't matter.

Contracts like NEX should be deployed with this `Payable` flag, because they always need to be payable.

**Comment by erikzhang** [comment]
> One last question. Does this affect the sending of NEP5 tokens to the contract, or the sending of NEO/GAS?

That's what I hesitated about. What do you think?

**Comment by localhuman** [comment]
I think it makes most sense in regards to NEO/GAS.  Contracts could enforce on their own whether they can receive NEP5 tokens.

**Comment by erikzhang** [comment]
@localhuman Is it better now?

**Comment by dicarlo2** [comment]
@erikzhang This will break every ongoing ICO when it's deployed. It will also break the Switcheo exchange contract. We should modify the changes in [1] to make sure it doesn't break existing contracts before deploying this.

[1] https://github.com/neo-project/neo/pull/233/commits/36896060b76c2cf290cd3bd2ff7a2c1264ba0d46

**Comment by erikzhang** [comment]
Don't worry, I will make new API to help those contracts to change their flags without re-deploy.

**Comment by dicarlo2** [comment]
@erikzhang Sounds good. I suppose that new API will cover my comment about NEP-7[1], right? When you make the new API please consider how alternate implementations (neo-python, neo-one) will handle it as well. For example, I'd like to make the changes to neo-one to support `payable` and `VerificationR` + the new API and have it running against the mainnet before the consensus nodes are updated so that there are no inconsistencies, if possible. However, since this only affects consensus, I suppose I can make the changes and deploy, neo-one nodes just won't relay any transactions that fail to be verified.

[1] https://github.com/neo-project/neo/pull/172#issuecomment-389371279

File: neo/Core/Transaction.cs [hunk 1]
```diff
@@ -385,6 +385,7 @@ private bool VerifyReceivingScripts()
             {
                 ContractState contract = Blockchain.Default.GetContract(hash);
                 if (contract == null) continue;
+                if (!contract.Payable) return false;
```
Comment (jsolman) [inline neo/Core/Transaction.cs]: Can we make this also check block height greater than activation block?

**Pull request merged**